rftp/server2.go

```diff
@@ -34,23 +34,35 @@ type clientConnection struct {
 func (c *clientConnection) writeResponse() {
 	log.Println("start writing response packets")
 	lastAck := uint8(0)
+	rateControl := &aimd{congRate: 4}
+	rateControl.start()
+	defer rateControl.stop()
 	for {
+		// TODO: Wait for rateControl.available() to be true. How should the caller
+		// of rateControl be informed? ACKs should be handled even if
+		// !rateControl.available().
 		var err error
 		select {
-
 		case md := <-c.metadata:
 			log.Printf("sending metadata for file %v: status: %v, size: %v, checksum: %x\n", md.fileIndex, md.status, md.size, md.checkSum)
 			md.ackNum = lastAck
 			err = sendTo(c.socket, *md)
+			if err != nil {
+				rateControl.onSend()
```

Comment:
Well, if sendTo returns an error, no packet was sent, right? But since we have bigger problems in that case, a bit less code complexity (in other words executing `rateControl.onSend()` no matter what) might be worth it.

Reply:
Yes, I guess most likely if it returns an error, no packet was sent not a 100% sure about the underlying syscall though.. And in that case i see, why you don't want to call `onSend`, but isn't this actually the opposite? It calls `onSend`, only when an error was returned and thus no packet was send, doesn't it?